test/ex_oauth2_provider/redirect_uri_test.exs

```diff
@@ -33,6 +33,26 @@ defmodule ExOauth2Provider.RedirectURITest do
     assert RedirectURI.validate("http://app.co/", []) == {:error, "Redirect URI must be an HTTPS/SSL URI"}
   end
 
+  test "validate http scheme when not required" do
+    uri = "http://app.co/"
+    assert RedirectURI.validate(uri, [force_ssl_in_redirect_uri: false]) == {:ok, uri}
+  end
+
+  test "validate custom url scheme" do
+    # RFC Spec - OAuth 2.0 for Native Apps
+    # https://tools.ietf.org/html/rfc8252#section-7.1
+    uri = "com.example.app:/oauth2redirect/example-provider"
+    assert RedirectURI.validate(uri, []) == {:ok, uri}
+    uri = "com.example.app://oauth2redirect/example-provider"
+    assert RedirectURI.validate(uri, []) == {:ok, uri}
+    uri = "com.example.app:/com.example.app/oauth2redirect/example-provider"
+    assert RedirectURI.validate(uri, []) == {:ok, uri}
+    uri = "com.example.app://com.example.app/oauth2redirect/example-provider"
+    assert RedirectURI.validate(uri, []) == {:ok, uri}
+    uri = "com.example.App://com.example.app/oauth2redirect/example-provider"
+    assert RedirectURI.validate(uri, []) == {:ok, uri}
```

Comment:
This assertion fails on elixir version 1.9.1 because it doesn't handle parsing uppercase schemes :(

Reply:
Good to know. I figured these tests are outside the scope of this module, so I removed them except the first. With the refactored logic it only tests whether the URI is absolute and has no fragments.